cpp/src/search/multisearch.cu

```diff
@@ -0,0 +1,604 @@
+    // Unfortunately, it's not constexpr until C++17, so we had to make a sort of a constexpr
+    // copy of it
+
+    std::pair<const E*, const E*> result =
+        cudf::util::equal_range(
+            column + valid_elements_subrange.start,
+            column + valid_elements_subrange.end_,
+            value
+        );
+
+    return index_range {
+        static_cast<gdf_size_type>(result.first - column),
+        static_cast<gdf_size_type>(result.second - column)
+    };
+}
+
+namespace detail {
+
+/**
+ * @brief A wrapper around @ref equality_subrange which obtains the
+ * (typed) needle record element.
+ *
+ * With C++17, this would have been a simple lambda within the `search` function.
+ * Although... we could have a version of type_dispatcher which imbues certain
+ * arguments with the column element type, e.g. based on some tag.
+ */
+struct equality_subrange_helper {
+
+    // Note: For some reason, we get an error message if we drop the __host__
+    // specifier - even though this function is never called by host-side code.
+    template <typename E>
+    __host__ __device__ index_range  operator()(
+        index_range                           range,
+        const void*            __restrict__   column,
+        const gdf_valid_type * __restrict__   column_validity,
+        const void*            __restrict__   values,
+        gdf_index_type                        value_index,
+        bool                                  value_is_valid,
+        bool                                  nulls_appear_before_values
+    )
+    {
+        return equality_subrange(
+            range,
+            static_cast<const E*>(column),
+            column_validity,
+            static_cast<const E*>(values)[value_index],
+            value_is_valid,
+            nulls_appear_before_values);
+    }
+};
+
+
+} // namespace detail
+
+struct search_result {
+    bool          found;
+    gdf_size_type pos;
+};
+
+search_result
+__device__
+search(
+    const void            * __restrict__ const * __restrict__  sorted_haystack_data,
+    const gdf_valid_type  * __restrict__ const * __restrict__  sorted_haystack_validities,
+    const void            * __restrict__ const * __restrict__  needle_data,
+    const gdf_valid_type  * __restrict__ const * __restrict__  needle_validities,
+    const gdf_dtype       * __restrict__                       column_data_types,
+    gdf_column_index_type                                      num_columns,
+    gdf_size_type                                              haystack_length,
+    gdf_size_type                                              needle_index,
+    bool                                                       nulls_appear_before_values,
+    bool                                                       find_first_greater
+)
+{
+    index_range potential_equality_range { 0, haystack_length };
+
+    for(gdf_column_index_type column_index = 0; column_index < num_columns; column_index++)
```

Comment:
Could we instead add comparison operators to `device_table_row_operators.cuh` for comparison of whole rows, rather than looping over columns here? e.g. the way that `rows_equal` works?